accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,547 @@
+
+## Supporting both runtime- and and static-reflection
+Previous discussions included the suggestion that both calling conventions and signature types are only exposed from the [`MetadataLoadContext`](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext) class since it is the preferred approach to read metadata (vs. runtime reflection) due to being agnostic to the runtime referenced by the various reflected types. However, this design proposes full fidelity between the runtime and `MetadataLoadContext` (other runtimes besides the CoreCLR, including Mono and AOT runtimes, will need to be updated of course).
+
+The reason to support calling conventions through `Type.GetFunctionPointerCallingConventions()` with the runtime are that the calling conventions may be useful in runtime-cases (not just introspection cases) and because there is a larger design factor with the CoreClr: the internal function pointer implementation uses the "CallKind" byte to hold the legacy calling conventions for things like determining if the VARARG convention is used. In addition, the CoreClr has a concept of "TypeKey" and "TypeHandle" and they are 1:1 today meaning that all `Type` instances are fully self-describing and thus, we can't have additional state hanging off of `Type` that is not part of the "key", such as calling conventions -- thus also the reason for bifurcation of signature types and non-signature types.
+
+This means that if we don't expose the calling conventions on `Type` for function pointers, some function pointer types will be castable and others not, and with no obvious way to inspect the `Type` to determine why. We could exposed just the "CallKind" byte on `Type`, but that is not desired since that will confuse and\or complicate the consumer's usage -- see "Normalization of calling conventions".
+
+Since we must already do the work of exposing the calling conventions including the newer ones that are read via custom modifiers, it makes sense then to expose all custom modifieres (via the new `Type.GetRequiredCustomModifiers()` and `Type.GetOptionalCustomModifiers()`.
+
+## Type identity 
+### Using of void*
+A function pointer addresscan be cast to any function pointer type when using `void*`:
+```cs
+delegate*<string> fn_string = &StringMethod;
+int i = ((delegate*<int>)(void*)fn_string)(); // Unsafe code allows any cast once void* is used
+static string StringMethod() => "Forty-Two";
+```
+And the IL just does a `ldftn` with no check:
```

Comment:
Right, IL does not care about type identify of pointers or function pointers. There is one type identify for all pointers and function pointers at IL level. Any (function) pointer is implicitly convertible to any other (function) pointer at IL level.

Reply:
I re-worded that a bit so it doesn't sound like an issue.